### PR TITLE
Add CodeCov for Ruby coverage reports

### PR DESCRIPTION
**Pull request by nschonni**
Not sure if this might overlap if the other Ruby one lands, but thought this might be an easier first step to see if the reporst are OK

### CODE REVIEW

**Comment by nschonni** [comment]
The comment on my test PR on my repo didn't have the right comment since it didn't have a baseline, but you can see the coverage report for the branch here https://app.codecov.io/gh/nschonni/mastodon/tree/jest-coverage

**Comment by nschonni** [comment]
Merged in the ruby coverage. You can see the test PR over here https://github.com/nschonni/mastodon/pull/117 with the details on https://app.codecov.io/gh/nschonni/mastodon/tree/codecov-actions

Looks like @mjankowski has added another 10% of the ruby since the last time I tested this out

**Comment by nschonni** [comment]
@Gargron @ClearlyClaire do you prefer not having the bot comment, and just having the status check? I think that might be configurable if the app is installed, with the comment suppressed in config

**Comment by renchap** [comment]
Do we really want codecov reports for the JS code?

We dont have any coverage yet, and I dont think we will get any, I would rather focus on end-to-end testing.

**Comment by nschonni** [comment]
There is some JS testing already like https://github.com/mastodon/mastodon/tree/main/app/javascript/mastodon/components/__tests__
E2E testing would be good, but unit testing covers other assumptions

**Comment by renchap** [comment]
I enabled the CodeCov app for the repo to get better statistics.

Some remarks:
- I am still unsure that the JS coverage reports will be useful. We do not have many JS tests, and the reports will not show anything helpful. I do not expect the coverage to increase this much, because the codebase is huge and not easily testable using Jest tests. The most efficient way to improve frontend testing will be end-to-end tests, which can not track JS coverage as far as I know
- We do not want this integration to be spammy. I saw that you can have multiple ways of reporting with Covecov, and I wonder which one (or ones) we want:
  - comment on the PR with the coverage. This will probably be too spammy?
  - Github status check with the coverage. What would be the criteria for the status check to be green?
  - Github PR annotations, maybe this could be another valid alternative?

**Comment by ClearlyClaire** [comment]
I agree that code coverage stats for the Javascript/Typescript part is probably not useful for the time being.

I think comments would probably be too spammy.

I think Github status checks may be the most useful, but I'm not sure what the criteria should be. Maybe we could put it into informational mode as a first step.

PR annotations/Github checks could also turn to be very useful, although I'm worried they could be a bit noisy. I guess we could turn them on and see how it looks.

**Comment by nschonni** [comment]
Should I delete the JS/TS CI stuff and the extra `jest:coverage` target?

Should I add a `codecov.yml` to track an of those threshold changes https://docs.codecov.com/docs/codecov-yaml

**Comment by ClearlyClaire** [comment]
> Should I delete the JS/TS CI stuff and the extra `jest:coverage` target?

Yes please!

> Should I add a `codecov.yml` to track an of those threshold changes https://docs.codecov.com/docs/codecov-yaml

I honestly don't know.

**Comment by nschonni** [comment]
Struggling with the Yarn 4 stuff, but I got it rebased without the JS parts

**Review by renchap — APPROVED** [review]
Let's do this, but we will probably need to configure it so it does not spam the PRs.

Will adjust the config once we have a few examples of how it performs.

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -2,13 +2,21 @@
 
 if ENV['DISABLE_SIMPLECOV'] != 'true'
   require 'simplecov'
+  require 'simplecov-lcov'
+  SimpleCov::Formatter::LcovFormatter.config.report_with_single_file = true
+  SimpleCov.formatter = SimpleCov::Formatter::LcovFormatter
   SimpleCov.start 'rails' do
+    enable_coverage :branch
+    enable_coverage_for_eval
```
Comment (mjankowski) [inline spec/spec_helper.rb:10]: Enabling this eval coverage causes some (not fixable) error output when running - https://github.com/simplecov-ruby/simplecov/issues/1057 -- the report stays accurate, but there's warnings that you can't actually do anything about (as described in the issue: from erb/haml/etc having an intermediate step in their template processes)

**Comment by mjankowski** [comment]
This has been enabled for ~3 months, do we find it useful?

For me - having the codecov reports *available* to look at on their site has been useful a few times. Having it add the comment on PRs (which creates an email as well) ... very much less so. I basically just delete these.

I'd probably be in favor of keeping the status check and ability to review on their site/app, but turning off the commenting, if that's possible.

**Comment by renchap** [comment]
Agreed, the comments are spammy, especially as you get an email notification for each.

According to the config, you can disable the comments, feel free to open a PR to do this!

**Pull request merged**